### PR TITLE
1824 add solution comparison tool

### PR DESCRIPTION
**Pull request by jonmaddock**
Add the ability to plot multiple solution vectors, normalised or unnormalised, with objective functions and RMS errors.

### CODE REVIEW

**Comment by jonmaddock** [comment]
It would be better to test this notebook, but also notebooks generally in Process, using the tools in #3091.

**Comment by jonmaddock** [comment]
@timothy-nunn you reviewed a previous version of this for me [here](https://git.ccfe.ac.uk/process/process/-/merge_requests/883), but I've made quite a few changes that I think improve on that and go beyond it. I'd be grateful for your review, thanks.

**Review by timothy-nunn — CHANGES_REQUESTED** [review]
Hi Jon, looks good other than a couple typing mistakes, possible (very unlikely) bugs, and a bug with the figure annotation. 

The only thing I wonder about is if we really want to have an implicit `normalising_tag` if `normalise=True` but its not provided? I lean towards explicit in cases like this and the behaviour would be easier to understand if we only normalise if a `normalising_tag` is provided. What are your thoughts on this?

File: process/io/plot_solutions.py [hunk 1]
```diff
@@ -0,0 +1,572 @@
...
+        plot_results_df,
+        opt_param_value_pattern=opt_param_value_pattern,
+        plot_title=plot_title,
+        normalise=normalise,
+        rmse_df=rmse_df,
+        normalising_tag=normalising_tag,
+        normalisation_type=normalisation_type,
+    )
+
+    # TODO Should return ax too?
```
[the remark this answers is not in the document]
Comment (jonmaddock) [reply process/io/plot_solutions.py:121]: `plot_solutions()` now returns `fig, df`.

File: process/io/plot_solutions.py [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+    non_normalising_solns_opt_params = _filter_opt_params(
+        non_normalising_solns, opt_param_value_pattern=opt_param_value_pattern
+    )
+
+    # Calculate the normalised differences between multiple solutions
+    # for optimisation parameters
+    normalising_soln_opt_params_np = normalising_soln_opt_params.to_numpy()
+    normalised_solns_opt_params = (
+        non_normalising_solns_opt_params - normalising_soln_opt_params_np
+    ) / abs(normalising_soln_opt_params_np)
```
[the remark this answers is not in the document]
Comment (jonmaddock) [reply process/io/plot_solutions.py:253]: Added a guard for this.

File: process/io/plot_solutions.py [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+    norm_objf_df = diffs_df.filter(regex=f"{NORM_OBJF_PATTERN}|{TAG}")
+
+    # Further separate objective function values only, for plotting
+    norm_objf_values_df = norm_objf_df.filter(regex=f"{NORM_OBJF_VALUE}|{TAG}")
+
+    # Acquire objective function name(s), then check only one type is being plotted
+    objf_list = norm_objf_df[NORM_OBJF_NAME].unique()
+
+    try:
+        assert len(objf_list) == 1
```
[the remark this answers is not in the document]
Comment (jonmaddock) [reply process/io/plot_solutions.py:340]: For some reason I forgot about that! Corrected.

**Comment by jonmaddock** [comment]
> Hi Jon, looks good other than a couple typing mistakes, possible (very unlikely) bugs, and a bug with the figure annotation.
> 
> The only thing I wonder about is if we really want to have an implicit `normalising_tag` if `normalise=True` but its not provided? I lean towards explicit in cases like this and the behaviour would be easier to understand if we only normalise if a `normalising_tag` is provided. What are your thoughts on this?

I agree; it would be much better to simplify and just have a `normalising_tag` argument: I'll change this.

File: process/io/plot_solutions.py [hunk 4]
```diff
@@ -0,0 +1,572 @@
...
+    # Adapt x axis label for normalisation type
+    if normalisation_type == "init":
+        x_axis_label = "Value normalised by initial value"
+    elif normalisation_type == "range":
+        x_axis_label = "Value normalised by parameter range"
+    else:
+        x_axis_label = "Value"
+
+    if normalise:
+        x_axis_label += f", normalised to {normalising_tag}"
```
[the remark this answers is not in the document]
Comment (jonmaddock) [reply process/io/plot_solutions.py:431]: Now if `normalising_tag` is `None`, this won't run.

File: process/io/plot_solutions.py [hunk 5]
```diff
@@ -0,0 +1,572 @@
...
+) -> pd.DataFrame:
+    """Plot multiple solutions, optionally normalised by a given solution.
+
+    :param runs_metadata: list of RunMetadata objects
+    :type runs_metadata: Sequence[RunMetadata]
+    :param plot_title: title of plot
+    :type plot_title: str
+    :param normalise: normalise optimisation parameters to a given solution, defaults to False
+    :type normalise: Optional[bool], optional
+    :param normalising_tag: tag for solution to normalise with, defaults to None
```
[the remark this answers is not in the document]
Comment (jonmaddock) [reply process/io/plot_solutions.py:66]: As per your general comment, I've removed the `normalise` argument to avoid this confusion.

**Comment by jonmaddock** [comment]
Hi Tim, thanks for you patience with this; this is ready for review again.

**Comment by jonmaddock** [comment]
Apologies, fixing tests.

**Comment by jonmaddock** [comment]
Ready for review again Tim. Regression tests are failing due to adding `objf_name` to the MFILE output, which isn't in the reference MFILE.

File: process/io/plot_solutions.py [hunk 6]
```diff
@@ -0,0 +1,572 @@
...
+    mfile = MFile(str(mfile_path))
+    mfile_data = {}
+
+    for var in mfile.data.keys():
+        if var.startswith("itvar"):
+            # Iteration variable: get name too
+            mfile_data[f"{var}_name"] = mfile.data[var].var_description
+
+        # Extract value
+        mfile_data[var] = mfile.data[var].get_scan(-1)
```
[the remark this answers is not in the document]
Comment (timothy-nunn) [reply process/io/plot_solutions.py:151]: This is now done at the top

**Review by timothy-nunn — APPROVED** [review]
Happy with these changes, especially the simplification of turning on/off normalisation. Cheers Jon.

**Pull request merged**